File: python/rosalind_iprb.py
from asyncore import read
from typing import Tuple

def read_data(filepath: str) -> Tuple[int, int, int]:
    with open(filepath, mode='r') as file:
        line = file.read().rstrip('\n')
        return tuple(map(int, line.split()))

def write_data(filepath: str, data: float) -> None:
    with open(filepath, mode='w') as file:
        file.write(str(data))

def dominant_allele_probability(homozygous_dominant: int, heterozygous_recessive: int, homozygous_recessive: int) -> float:
    k = homozygous_dominant
    m = heterozygous_recessive
    n = homozygous_recessive

    t = k + m + n
    
    pk = k / t
    pm = m / t
    pn = n / t

    # homozygous recessive & homozygous recessive
    c1 = pn * ((n - 1) / (t - 1))

    # homozygous recessive & heterozygous recessive
    c2 = 2 * pn * (m / (t - 1)) * 0.5

    # heterozygous recessive & heterozygous recessive
    c3 = pm * ((m - 1) / (t - 1)) * 0.25

    return 1 - (c1 + c2 + c3)

if __name__ == '__main__':
    data = read_data('input/rosalind_iprb.txt')
    result = dominant_allele_probability(*data)
    write_data('output/rosalind_iprb.txt', result)

File: java/RosalindIPRB.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;

public class RosalindIPRB {
    private static int[] readData(String filePath) throws IOException {
        Path inputFile = Paths.get(filePath);
        String inputLine = Files.readAllLines(inputFile).get(0);
        return Arrays.stream(inputLine.split(" ")).mapToInt(Integer::parseInt).toArray();
    }

    private static void writeData(String filePath, double data) throws IOException {
        Path outputFile = Paths.get(filePath);

        Files.deleteIfExists(outputFile);
        Files.createDirectories(outputFile.getParent());
        Files.createFile(outputFile);

        Files.writeString(outputFile,String.valueOf(data));
    }

    // k - homozygous dominant
    // m - heterozygous recessive
    // n - homozygous recessive
    private static double dominantAlleleProbability(int k, int m, int n) {
        int t = k + m + n;

        double pk = (double)k / t;
        double pm = (double)m / t;
        double pn = (double)n / t;

        // homozygous recessive & homozygous recessive
        double c1 = pn * ((double)(n - 1) / (double)(t - 1));

        // homozygous recessive & heterozygous recessive
        double c2 = 2 * pn * ((double)m / (double)(t - 1)) * 0.5;

        // heterozygous recessive & heterozygous recessive
        double c3 = pm * ((double)(m - 1) / (double)(t - 1)) * 0.25;

        return 1 - (c1 + c2 + c3);
    }

    public static void main(String[] args) throws IOException {
        int[] data = readData("input/rosalind_iprb.txt");
        double result = dominantAlleleProbability(data[0], data[1], data[2]);
        writeData("output/rosalind_iprb.txt", result);
    }
}
